tstdc/src/mutex.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! Mutex functionality for an SGX enclave
+
+use core::cell::UnsafeCell;
+use mc_sgx_tstdc_sys::{
+    sgx_thread_mutex_destroy, sgx_thread_mutex_lock, sgx_thread_mutex_trylock,
+    sgx_thread_mutex_unlock,
+};
+use mc_sgx_tstdc_sys_types::{sgx_thread_mutex_t, SGX_THREAD_NONRECURSIVE_MUTEX_INITIALIZER};
+
+/// Errors when interacting with [`Mutex`]es
+#[derive(Copy, Clone, Debug, displaydoc::Display, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub enum Error {
+    /// Invalid operation on the mutex
+    Invalid,
+    /// Mutex is currently locked by another thread
+    Busy,
+}
+
+type Result<T> = core::result::Result<T, Error>;
+
+/// A mutex inside of an SGX enclave
+///
+/// NB: per the documentation of
+/// [`sgx_thread_mutex_lock()`](https://download.01.org/intel-sgx/sgx-linux/2.18/docs/Intel_SGX_Developer_Reference_Linux_2.18_Open_Source.pdf#%5B%7B%22num%22%3A303%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C94.5%2C341.25%2C0%5D)
+/// a mutex should not be locked across root ECALLs.
+#[derive(Debug, Default)]
+pub struct Mutex(UnsafeCell<sgx_thread_mutex_t>);
+
+// SAFETY: The `sgx_thread_mutex_*` C functions utilize a spinlock to prevent
```

Comment:
Should this `SAFETY` block go above the struct definition on 29? I thought that's where they normally go but might be wrong.

Reply:
Moved block to before struct.
I think in this case when there are multiple unsafe blocks it makes more sense.
The alternative is to document each unsafe block but they all seem to be for the same reason.